### PR TITLE
US102307: Added swagger annotations for ACS Connectors APIs.

### PR DESCRIPTION
**Pull request by irinaepshteyn**

### CODE REVIEW

File: model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java [hunk 1]
```diff
@@ -32,6 +36,7 @@ public AttributeAdapterConnection(final AttributeAdapterConnection other) {
         this.uaaClientSecret = other.uaaClientSecret;
     }
 
+    @ApiModelProperty(value = "URL to a remote attribute store adapter", required = true)
```
Comment (anubhavi25) [inline model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Might be more clearly worded as "URL to retrieve attributes from"?

File: model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java [hunk 2]
```diff
@@ -48,6 +54,9 @@ public void setUaaTokenUrl(final String uaaTokenUrl) {
         this.uaaTokenUrl = uaaTokenUrl;
     }
 
+    @ApiModelProperty(
+            value = "An OAuth client that coud be used to obtain an access token for this adapter",
```
Comment (anubhavi25) [inline model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Typo: coud -> could. Also might be more clearly worded as "OAuth client used to obtain an access token for this adapter"?

File: model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java [hunk 3]
```diff
@@ -56,6 +65,7 @@ public void setUaaClientId(final String uaaClientId) {
         this.uaaClientId = uaaClientId;
     }
 
+    @ApiModelProperty(value = "A secret for an OAuth client for this adapter", required = true)
```
Comment (anubhavi25) [inline model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Might be more clearly worded as "OAuth client secret used to obtain an access token for this adapter"?

File: model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java [hunk 4]
```diff
@@ -8,14 +8,19 @@
 
 import com.fasterxml.jackson.annotation.JsonProperty;
 
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(description = "Connector configuration for external resource or subject atributes.")
```
Comment (anubhavi25) [inline model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java]: Typo: atributes -> attributes

File: model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java [hunk 5]
```diff
@@ -8,14 +8,19 @@
...
 public class AttributeConnector {
-    private boolean isActive;
+    private boolean isActive = false;
 
     @JsonProperty(required = false)
     private int maxCachedIntervalMinutes = 480; //default value
 
     private Set<AttributeAdapterConnection> adapters;
 
+    @ApiModelProperty(value = "A flag to enable or disable attribute connector. Disabled by default")
```
Comment (anubhavi25) [inline model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java]: Might be more clearly worded as "A flag to enable or disable the retrieval of remote attributes. Disabled by default."?

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 6]
```diff
@@ -15,24 +15,42 @@
...
+
+
 @RestController
 public class AttributeConnectorController extends BaseRestApi {
 
     @Autowired
     private AttributeConnectorServiceImpl service;
 
+    @ApiOperation(
+            value = "Creates or updates connector configururation for external resource attributes for the given zone.",
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Typo: configururation -> configuration

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 7]
```diff
@@ -15,24 +15,42 @@
...
+            tags = { "Attribute Connector Management" },
+            response = AttributeConnector.class)
+    @ApiResponses(
+            value = { @ApiResponse(
+                    code = 201,
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Is it possible to mention all possible response codes/descriptions here (i.e. `200` for updates and `201` for creates)?

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 8]
```diff
@@ -15,24 +15,42 @@
...
+                    message = "Connector configuration for the given zone is successfully created.") })
     @RequestMapping(
             method = PUT,
             value = V1 + AcsApiUriTemplates.RESOURCE_CONNECTOR_URL,
-            consumes = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity<AttributeConnector> putResourceConnector(@RequestBody final AttributeConnector connector) {
+            consumes = MediaType.APPLICATION_JSON_VALUE,
+            produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<AttributeConnector> putResourceConnector(@ApiParam(
+            value = "New or updated connector configururation for external resource attributes",
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Typo: configururation -> configuration

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 9]
```diff
@@ -15,24 +15,42 @@
...
+            value = { @ApiResponse(
+                    code = 201,
+                    message = "Connector configuration for the given zone is successfully created.") })
     @RequestMapping(
             method = PUT,
             value = V1 + AcsApiUriTemplates.RESOURCE_CONNECTOR_URL,
-            consumes = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity<AttributeConnector> putResourceConnector(@RequestBody final AttributeConnector connector) {
+            consumes = MediaType.APPLICATION_JSON_VALUE,
+            produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: I thought we return an empty body for upserts in which case this isn't needed, no?

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 10]
```diff
@@ -47,6 +65,12 @@
         }
     }
 
+    @ApiOperation(
+            value = "Retrieves connector configururation for external resource attributes for the given zone.",
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Typo: configururation -> configuration

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 11]
```diff
@@ -47,6 +65,12 @@
...
+            tags = { "Attribute Connector Management" },
+            response = AttributeConnector.class)
+    @ApiResponses(
+            value = { @ApiResponse(code = 404, message = "Connector configuration for the given zone is not found.") })
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: See comment above re. mentioning all possible response codes.

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 12]
```diff
@@ -63,6 +87,16 @@
         }
     }
 
+    @ApiOperation(
+            value = "Deletes connector configururation for external resource attributes for the given zone.",
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Typo: configururation -> configuration

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 13]
```diff
@@ -76,11 +110,22 @@
         }
     }
 
+    @ApiOperation(
+            value = "Creates or updates connector configururation for external subject attributes for the given zone.",
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Typo: configururation -> configuration

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 14]
```diff
@@ -95,6 +140,12 @@
         }
     }
 
+    @ApiOperation(
+            value = "Retrieves connector configururation for external subject attributes for the given zone.",
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Typo: configururation -> configuration

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 15]
```diff
@@ -111,6 +162,16 @@
         }
     }
 
+    @ApiOperation(
+            value = "Deletes connector configururation for external subject attributes for the given zone.",
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Typo: configururation -> configuration

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 16]
```diff
@@ -95,6 +140,12 @@
         }
     }
 
+    @ApiOperation(
+            value = "Retrieves connector configururation for external subject attributes for the given zone.",
+            tags = { "Attribute Connector Management" },
+            response = AttributeConnector.class)
+    @ApiResponses(
+            value = { @ApiResponse(code = 404, message = "Connector configuration for the given zone is not found.") })
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: See comment above re. mentioning all possible response codes.

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 17]
```diff
@@ -76,11 +110,22 @@
         }
     }
 
+    @ApiOperation(
+            value = "Creates or updates connector configururation for external subject attributes for the given zone.",
+            tags = { "Attribute Connector Management" },
+            response = AttributeConnector.class)
+    @ApiResponses(
+            value = { @ApiResponse(
+                    code = 201,
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: See comment above re. mentioning all possible response codes.

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 18]
```diff
@@ -76,11 +110,22 @@
...
+                    message = "Connector configuration for the given zone is successfully created.") })
     @RequestMapping(
             method = PUT,
             value = V1 + AcsApiUriTemplates.SUBJECT_CONNECTOR_URL,
-            consumes = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity<AttributeConnector> putSubjectConnector(@RequestBody final AttributeConnector connector) {
+            consumes = MediaType.APPLICATION_JSON_VALUE,
+            produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: See comment above re. redundant `produces` field

File: model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java [hunk 19]
```diff
@@ -32,6 +39,7 @@ public void setMaxCachedIntervalMinutes(final int maxCachedIntervalMinutes) {
         this.maxCachedIntervalMinutes = maxCachedIntervalMinutes;
     }
 
+    @ApiModelProperty(value = "A set of attribute store adapters associated with this connector", required = true)
```
Comment (anubhavi25) [inline model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java]: Might be more clearly worded as "A set of adapters used to retrieve attributes from"? Also should we mention here or in a controller-related Swagger annotation that only one adapter is currently supported?

**Review by anubhavi25 — CHANGES_REQUESTED** [review]

File: model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java (hunk 1 above)
Comment (irinaepshteyn) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Then it sounds like it's a URL to attribute store itself.  How about "Adapter URL to retrieve attributes from"?

File: model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java (hunk 2 above)
Comment (irinaepshteyn) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Agreed

File: model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java (hunk 3 above)
Comment (irinaepshteyn) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Agreed

File: model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java (hunk 4 above)
Comment (irinaepshteyn) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java]: Thanks.

File: model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java (hunk 5 above)
Comment (irinaepshteyn) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java]: Agreed

File: model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java (hunk 19 above)
Comment (irinaepshteyn) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeConnector.java]: Agreed.

File: model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java (hunk 1 above)
Comment (anubhavi25) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Works 👍

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java (hunk 7 above)
Comment (irinaepshteyn) [reply service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: 200, 401, 403 are displayed by swagger by default without special annotations.

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java (hunk 18 above)
Comment (irinaepshteyn) [reply service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Good catch.  I changed return type from ResponseEntity<AttributeConnector> to ResponseEntity<?>

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java (hunk 7 above)
Comment (anubhavi25) [reply service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: At least for `200` shouldn't we override the message with something connector-specific like "Connector configuration for the given zone was successfully updated."?

Comment (irinaepshteyn) [reply service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: I couldn't find a way to display a description for the default response status.  Default status is always displayed at the top as "Response Class (Status 200)" or "Response Class (Status 204)" if I change the default to 204.

Version of swagger we use is 2.2.2.  May be it's possible in a later versions.

**Comment by irinaepshteyn** [comment]
Addressed pull request comments.  Please review

File: model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java (hunk 1 above)
Comment (anubhavi25) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Did we decide to change this or leave it? Latest rebase (4a2f699) doesn't show this line being changed.

Comment (irinaepshteyn) [reply model/src/main/java/com/ge/predix/acs/rest/AttributeAdapterConnection.java]: Missed it.  Try again now.

File: service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java [hunk 20]
```diff
@@ -15,24 +15,40 @@
...
 
 import com.ge.predix.acs.commons.web.AcsApiUriTemplates;
 import com.ge.predix.acs.commons.web.BaseRestApi;
 import com.ge.predix.acs.commons.web.RestApiException;
 import com.ge.predix.acs.rest.AttributeConnector;
 
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiParam;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;;
```
Comment (anubhavi25) [inline service/src/main/java/com/ge/predix/acs/attribute/connector/management/AttributeConnectorController.java]: Remove extra semicolon?

**Review by anubhavi25 — APPROVED** [review]

**Comment by irinaepshteyn** [comment]
Tests PASSed:
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-build-pullrequest/124/

https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/389/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/390/
https://predix1.jenkins.build.ge.com/job/Predix-Security/job/ACS/job/ACS-PullRequest/job/acs-integration-pullrequest/393/

**Comment by anubhavi25** [comment]
@irinaepshteyn please merge once PR integration tests pass

**Pull request merged**